Automattic/wp-calypso
# Store Promotions: Make proper promotions objects

**NOTE: This PR will not be merged. Instead I will merge the next version of this code. Please review #19092 instead as any changes will need to go there.**

(I'm keeping this PR around until #19092 is merged as the commits for that PR are still based on the commits here.)

This is a refactor of the previous code which tried to aggregate coupons
and products into promotions. This system wasn't working because the
abstraction broke down every time. Instead, this approach seeks to
derive proper promotions objects starting from a coupon or a product,
and then will eventually save down to those respective objects. But
while the client is dealing with the promotions, they will be homogenous
objects.

This approach has some UX benefits as well as benefits for upgrading to
a Promotions API implementation if/when that becomes available.

NOTE: This is a branch off of earlier PRs and will invalidate some other PRs as it goes. The chain of PRs forks here. More to follow.

To Test:
1. `npm test`
2. `npm start`
3. Visit `http://calypso.localhost:3000/store/promotion/<site url>`
4. Change the promotion type selection and observe the fields below changing.
5. Attempt to enter text into the fields below.
6. Attempt to enter an invalid percent.
7. Attempt to enter a negative currency amount.

Overall, I'm much more pleased with this approach from both a code logic standpoint, and also what it enables for the UX (like keeping values between switching types, etc.)

![new_promotion_ _coderkevin_ _wordpress_com](https://user-images.githubusercontent.com/945228/31925124-2ef31544-b84c-11e7-90f3-1fdb59ceef03.png)

![new_promotion_ _coderkevin_ _wordpress_com](https://user-images.githubusercontent.com/945228/31925138-45dec5a0-b84c-11e7-83da-dafc32b1b3de.png)

![new_promotion_ _coderkevin_ _wordpress_com](https://user-images.githubusercontent.com/945228/31925147-50ef11c0-b84c-11e7-9a55-7ff5f83d3b41.png)
